topology/pynest/tests/test_spatial_kernels.py

```diff
@@ -0,0 +1,811 @@
+        '''
+        Construct a test object.
+
+        Parameters
+        ----------
+            seed         : Random seed for test
+            dim          : Dimensions (2 or 3)
+            L: Side length of area / volume.
+            N: Number of nodes.
+            open_bc      : Network with open boundary conditions
+            x0, y0       : Location of source neuron; open_bc only
+        '''
+
+        if dim != 2 and open_bc:
+            raise ValueError("open_bc only supported for 2D")
+
+        self._seed = seed
+        self._dimensions = dim
+        self._L = float(L)
+        self._N = N
+        self._open_bc = open_bc
+        self._x_d, self._y_d = x0, y0
+
+        if (self._dimensions == 2):
+            if (self._open_bc):
+                self._max_dist = self._L * np.sqrt(2)
+            else:
+                self._max_dist = self._L / np.sqrt(2)
+        elif (self._dimensions == 3):
+            self._max_dist = self._L * np.sqrt(3) / 2
+
+        self._target_dists = None
+        self._all_dists = None
+
+    def _create_dist_data(self):
+
+        self._reset(self._seed)
+        self._build()
+        self._connect()
+
+        self._target_dists = sorted(self._target_distances())
+        self._all_dists = self._all_distances()
+
+    def _reset(self, seed):
+        '''Reset simulator and seed PRNGs.'''
+
+        pass
+
+    def _build(self):
+        '''Create populations.'''
+
+        raise NotImplementedError()
+
+    def _connect(self):
+        '''Connect populations.'''
+
+        raise NotImplementedError()
+
+    def _kernel(self, D):
+        '''Distance dependent probability function (kernel).'''
+
+        raise NotImplementedError()
+
+    def _positions(self):
+        '''Return list of position tuples for all nodes.'''
+
+        raise NotImplementedError()
+
+    def _target_positions(self):
+        '''Return list of position tuples of all connected target nodes.'''
+
+        raise NotImplementedError()
+
+    def _all_distances(self):
+        '''Return list with distances to all nodes.'''
+
+        raise NotImplementedError()
+
+    def _target_distances(self):
+        '''
+        Return list with distances from source node to all connected
+        target nodes.
+        '''
+
+        raise NotImplementedError()
+
+    def _roi_2d(self, x, y, L):
+        '''
+        Moves coordinates (x,y) to triangle area (x',y') in [0,L/2]X[0,x']
+        without loss of generality
+        '''
+        x = -x if (x >= -self._L / 2.) & (x < 0) else x
+        y = -y if (y >= -self._L / 2.) & (y < 0) else y
+        return np.array([x, y]) if x > y else np.array([y, x])
+
+    def _pdf(self, D):
+        '''
+        Unnormalized probability density function (PDF).
+
+        Parameters
+        ----------
+            D: Distance in interval [0, max_dist].
+
+        Return values
+        -------------
+            Unnormalized PDF at distance D.
+        '''
+
```

Comment:
This function is quite long and cluttered. I had a very quick look in the thesis, but when I couldn't find a total definition of the PDF I gave up. Looking at this function, it does make sense that the definition of the PDF is scattered or not written down explicitly in the thesis, as it has so many different constraints/conditions, but I have to admit that I have not looked at the math here in detail. Seems to be working though, as the test pass! Some comments might be helpful, but it might just make the function more dense.

Reply:
I don't have more details at the moment, but I will add documentation when it becomes available (as publication). I hope that is good enough for now.